src/vsr/journal.zig

```diff
@@ -221,73 +248,126 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
         }
 
         /// Asserts that headers are .reserved (zeroed) from `op_min` (inclusive).
-        pub fn assert_headers_reserved_from(self: *Self, op_min: u64) void {
+        pub fn assert_headers_reserved_from(self: *const Self, op_min: u64) void {
             // TODO Snapshots
-            for (self.headers[op_min..]) |header| assert(header.command == .reserved);
+            for (self.headers) |*header| {
+                assert(header.command == .reserved or header.op < op_min);
+            }
+        }
+
+        /// Returns whether this is a fresh database WAL; no prepares (except the init) have ever
+        /// been written.
+        pub fn is_empty(self: *const Self) bool {
+            if (self.headers[0].operation != .init) return false;
+
+            // If the message immediately following the initial prepare is damaged, we must fall
+            // back to VSR recovery protocol (i.e. treat this as a non-empty WAL) since that
+            // message may have been a prepare.
+            if (self.faulty.bit(Slot{ .index = 1 })) return false;
+
+            // Other headers are allowed to be damaged (they will be repaired from status=normal)
+            // but there shouldn't be any prepares.
+            for (self.headers[1..]) |*header| {
+                if (header.command == .prepare) return false;
+            }
+
+            return true;
+        }
+
+        pub fn slot_for_op(self: *const Self, op: u64) Slot {
+            return Slot{ .index = op % self.headers.len };
```

Comment:
Can we use just `slot_count` here since it's shorter and also comptime known? The compiler may be able to eek out a little more optimization, e.g. replace the mod with a bitwise mask.